blog/2018-07-17-UnitTestsInXslxFileForConferenceScheduling.adoc

```diff
@@ -0,0 +1,122 @@
+    scoreVerifier.assertHardWeight(ROOM_CONFLICT, 0, solution);
+    // talks in same room without overlapping time slots
+    talk1.withRoom(room1).withTimeslot(slot1);
+    talk2.withRoom(room1).withTimeslot(slot3);
+    scoreVerifier.assertHardWeight(ROOM_CONFLICT, 0, solution);
+    // talks in same room with overlapping time slots
+    talk2.withTimeslot(slot2);
+    scoreVerifier.assertHardWeight(ROOM_CONFLICT, -10, solution);
+    }
+----
+
+In order to test room conflict, you need to initialize two talks, three timeslots and one room.
+However, the previous snippet of code is too much for such a simple unit test, most of the boilerplate code is for
+initializing required fields for the conference solution that you do not need for the unit test, and you must do that for every single unit test.
+For more complex constraints, it gets too cumbersome to write traditional unit tests and reason about them.
+
+
+== Unit tests in .xlsx files
+In order to avoid initializing unwanted fields, you can take advantage of https://github.com/kiegroup/optaplanner/blob/master/optaplanner-examples/src/main/java/org/optaplanner/examples/conferencescheduling/persistence/ConferenceSchedulingXlsxFileIO.java[ConferenceSchedulingXlsxFileIO]
+to initialize them for you, and only write what you want to use. Also you can use them for all other unit tests.
+
+To test room conflict using xlsx file, what you want is three timeslots, two talks and one room:
+
+image::timeslots.png[link="timeslots.png" role="thumbnail"]
+image::talks.png[link="talks.png" role="thumbnail"]
+image::rooms.png[link="rooms.png" role="thumbnail"]
+
+After you initialize the required fields, you need to create a separate sheet for every score verification.
+
+image::roomConflict1.png[link="roomConflict1.png" role="thumbnail" alt="roomConflict1"]
+image::roomConflict2.png[link="roomConflict2.png" role="thumbnail"]
+
+In every test sheet (blue color), specify the constraint package, constraint name and expected score. Then assign the talks to
+rooms and timeslots and visualize them easily. Note that you do not need to list all the timeslots and rooms
+declared in Timeslots and Rooms sheets.
+
+
+== Conclusion
+To test your score rules in an xlsx file, follow these steps:
+
+1. List all the required fields for your tests in the setup sheets.
+2. For every score verification, create a separate blue test sheet with the constraint package, constraint name and expcted score.
+3. List only the fields that you want to use for the corresponding rule.
+4. Set _testFileName_ in https://github.com/kiegroup/optaplanner/blob/master/optaplanner-examples/src/main/test/java/org/optaplanner/examples/conferencescheduling/solver/ConferenceSchedulingScoreRulesXlsxTest.java[ConferenceSchedulingScoreRulesXlsxTest] and run the test file.
+
+The test file clones a separate solution instance for every test sheet, so the assigned planning variables in one sheet will not affect
```

Comment:
that's an implementation detail. Those don't belong in the conclusion.

An Introduction should specify the problem, incite people to read the article
A conclusion should reiterate and summarize the solution, the gain.